Makefile

```diff
@@ -87,14 +87,27 @@ authors: ## generate AUTHORS file from git history
 	scripts/docs/generate-authors.sh
 
 .PHONY: completion
-completion: binary
-completion: /etc/bash_completion.d/docker
-completion: ## generate and install the completion scripts
-
-.PHONY: /etc/bash_completion.d/docker
-/etc/bash_completion.d/docker: ## generate and install the bash-completion script
-	mkdir -p /etc/bash_completion.d
-	docker completion bash > /etc/bash_completion.d/docker
+completion: shell-completion
+completion: ## generate and install the shell-completion scripts
+# Note: this uses system-wide paths, and so may overwrite completion
+# scripts installed as part of deb/rpm packages.
+#
+# Given that this target is intended to debug/test updated versions, we could
+# consider installing in per-user (~/.config, XDG_DATA_DIR) paths instead, but
+# this will add more complexity.
+#
+# See https://github.com/docker/cli/pull/5770#discussion_r1927772710
+	install -D -p -m 0644 ./build/completion/bash/docker /usr/share/bash-completion/completions/docker
+	install -D -p -m 0644 ./build/completion/fish/docker.fish debian/docker-ce-cli/usr/share/fish/vendor_completions.d/docker.fish
+	install -D -p -m 0644 ./build/completion/zsh/_docker debian/docker-ce-cli/usr/share/zsh/vendor-completions/_docker
+
+.PHONY: shell-completion
+shell-completion: ## generate shell-completion scripts
+# requires either "binary" or "dynbinary" to be built.
+# but only build if it's not there, to prevent existing binaries
+# from being replaced.
+	@ [ -f build/docker ] || $(MAKE) binary
+	@ ./scripts/build/completion
```

Comment:
Slightly hacky here; the `make binary` and `make dynbinary` targets overwrite the `build/binary` file; I'm  trying to avoid cases where either a Dynamic or a Static binary was built, and generating the completion-scripts now replacing the binary.